client/pyscripts/fm11rf08_full.py

```diff
@@ -0,0 +1,942 @@
+def decodeBlock0():
+	global uid
+	global uids
+
+	# We do this early so we can name the logfile!
+	uids = blk0[0:11]                            # UID string  : "11 22 33 44"
+	uid  = int(uids.replace(' ', ''), 16)        # UID (value) : 0x11223344
+	startlog(uid, append=False)
+
+	lprint(prompt)
+	lprint(f"{prompt}              UID         BCC         ++----- RF08 ID -----++")
+	lprint(f"{prompt}              !           !  SAK      !!                   !!")
+	lprint(f"{prompt}              !           !  !  ATQA  !!     RF08 Hash     !!")
+	lprint(f"{prompt}              !---------. !. !. !---. VV .---------------. VV")
+	#                              0           12 15 18    24 27                45
+	#                              !           !  !  !     !  !                 !
+	#                              00 11 22 33 44 55 66 77 88 99 AA BB CC DD EE FF
+	lprint(f"{prompt}   Block 0  : {blk0}")
+
+	# --- decode block 0 ---
+
+	bcc  = int(blk0[12:14], 16)                  # BCC
+	chk  = 0                                     # calculate checksum
+	for h in uids.split():
+		chk ^= int(h, 16)
+
+	sak  = int(blk0[15:17], 16)                  # SAK
+	atqa = int(blk0[18:23].replace(' ',''), 16)  # 0x7788
+
+	fida = int(blk0[24:26], 16)                  # Fudan ID 0x88
+	fidb = int(blk0[45:47], 16)                  # Fudan ID 0xFF
+	fid  = (fida<<8)|fidb                        # Fudan ID 0x88FF
+
+	hash = blk0[27:44]                           # Fudan hash "99 AA BB CC DD EE"
+
+	type = f"[{fida:02X}:{fidb:02X}]"            # type/name
+	if fidb == 0x90:
+		if fida == 0x01 or fida == 0x03 or fida == 0x04:
+			type += " - Fudan FM11RF08S"
+
+	elif fidb == 0x1D:
+		if fida == 0x01 or fida == 0x02 or fida == 0x03:
+			type += " - Fudan FM11RF08"
+
+	elif fidb == 0x91 or fidb == 0x98:
+		type += " - Fudan FM11RF08 (never seen in the wild)"
+
+	else:
+		type += " - Unknown (please report)"
+
+	# --- show results ---
+
+	lprint(prompt)
+
+	lprint(f"{prompt}   UID/BCC  : {uid:08X}/{bcc:02X} - ", end='')
+	if bcc == chk:  lprint("verified")
+	else:           lprint(f"fail. Expected {chk:02X}")
+
+	lprint(f"{prompt}   SAK      : {sak:02X} - ", end='')
+	if   sak == 0x01:  lprint("NXP MIFARE TNP3xxx 1K")
+	elif sak == 0x08:  lprint("NXP MIFARE CLASSIC 1k | Plus 1k | Ev1 1K")
+	elif sak == 0x09:  lprint("NXP MIFARE Mini 0.3k")
+	elif sak == 0x10:  lprint("NXP MIFARE Plus 2k")
+	elif sak == 0x18:  lprint("NXP MIFARE Classic 4k | Plus 4k | Ev1 4k")
+	else:              lprint("{unknown}")
+
+	lprint(f"{prompt}   ATQA     : {atqa:04X}")   # show ATQA
+	lprint(f"{prompt}   Fudan ID : {type}")       # show type
+	lprint(f"{prompt}   Fudan Sig: {hash}")       # show ?Partial HMAC?
+	lprint(f"{prompt}   Dark Key : {dkey}")       # show key
+
+#+=============================================================================
+# Fudan validation
+# >> "blk0"
+#==============================================================================
+def fudanValidate():
```

Comment:
it would be interesting to isolate this function in a separate file to be imported, so other ppl can use it in their scripts s well, or call the script with a provided block0

Reply:
qv. https://github.com/RfidResearchGroup/proxmark3/pull/2613#discussion_r1830270895